src/core/experimental/array.d

```diff
@@ -0,0 +1,2299 @@
+    {
+        auto a = Array!int(1, 2, 3);
+        assert(a.front == 1);
+        a.front = 0;
+        assert(a.front == 0);
+    }
+
+    /**
+     * Advance to the next element in the array. The user must check
+     * that the array isn't `empty`, prior to calling this function.
+     *
+     * Complexity: $(BIGOH 1).
+     */
+    @nogc nothrow pure @safe
+    void popFront()
+    {
+        assert(!empty, "Array.popFront: Array is empty");
+        payload = payload[1 .. $];
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto stuff = [1, 2, 3];
+        auto a = Array!int(stuff);
+        size_t i = 0;
+        while (!a.empty)
+        {
+            assert(a.front == stuff[i++]);
+            a.popFront;
+        }
+        assert(a.empty);
+    }
+
+    /**
+     * Advance to the next element in the array. The user must check
+     * that the array isn't `empty`, prior to calling this function.
+     *
+     * This must be used in order to iterate through a `const` or `immutable`
+     * array For a mutable array this is equivalent to calling `popFront`.
+     *
+     * Returns:
+     *      an array that starts with the next element in the original array.
+     *
+     * Complexity: $(BIGOH 1).
+     */
+    Qualified tail(this Qualified)()
+    {
+        assert(!empty, "Array.tail: Array is empty");
+
+        static if (is(Qualified == immutable) || is(Qualified == const))
+        {
+            return this[1 .. $];
+        }
+        else
+        {
+            return .tail(this);
+        }
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto ia = immutable Array!int([1, 2, 3]);
+        assert(ia.tail.front == 2);
+    }
+
+    /**
+     * Eagerly iterate over each element in the array and call `fun` over each
+     * element. This should be used to iterate through `const` and `immutable`
+     * arrays.
+     *
+     * Normally, the entire array is iterated. If partial iteration (early stopping)
+     * is desired, `fun` needs to return a value of type `int` (`-1` to stop, or
+     * anything else to continue the iteration.
+     *
+     * Params:
+     *      fun = unary function to apply on each element of the array.
+     *
+     * Returns:
+     *      `true` if it has iterated through all the elements in the array, or
+     *      `false` otherwise.
+     *
+     * Complexity: $(BIGOH n).
+     */
+    template each(alias fun)
+    {
+        //import std.functional : unaryFun;
+
+        bool each(this Q)()
+        //if (is (typeof(unaryFun!fun(T.init))))
+        {
+            //alias fn = unaryFun!fun;
+            alias fn = fun;
+
+            // Iterate through the underlying payload
+            // The array is kept alive (rc > 0) from the caller scope
+            foreach (ref e; this.payload)
+            {
+                static if (!is(typeof(fn(T.init)) == int))
+                {
+                    cast(void) fn(e);
```

Comment:
For integration with `Flag!"each"`, I suggest:

```
static if (is(typeof(fn(T.init)) == int) Result)
{
    static if (is(typeof(Result.init == Result.init)) { if (fn(e) == Result.init) return false; }
    cast(void) fn(E);
}
```

Reply:
Not sure if this is correct, though, as a function that returns `bool` will now match the `Flag!each` pattern